toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/PopupState.kt

```diff
@@ -0,0 +1,358 @@
+     * @return returns the [PopupElementStateCollection] created.
+     */
+    internal fun createStates(
+        popup: Popup,
+        attachments: List<PopupAttachment>,
+        coroutineScope: CoroutineScope,
+        ignoreList: Set<Class<out PopupElement>> = emptySet(),
+    ): PopupElementStateCollection {
+        val states = mutablePopupElementStateCollection()
+        val elements: List<PopupElement> = popup.evaluatedElements
+        // Filter out elements that are part of the ignore list.
+        val filteredElements = elements.filter { element ->
+            !ignoreList.contains(element::class.java)
+        }
+        filteredElements.forEach { element ->
+            when (element) {
+                is TextPopupElement -> {
+                    states.add(
+                        element,
+                        TextElementState(element = element, popup = popup)
+                    )
+                }
+
+                is AttachmentsPopupElement -> {
+                    states.add(
+                        element,
+                        AttachmentsElementState(
+                            attachmentPopupElement = element,
+                            attachments = attachments
+                        )
+                    )
+                }
+
+                is FieldsPopupElement -> {
+                    states.add(
+                        element,
+                        FieldsElementState(element = element, popup = popup)
+                    )
+                }
+
+                is MediaPopupElement -> {
+                    states.add(
+                        element,
+                        MediaElementState(element = element, popup = popup)
+                    )
+                }
+
+                is UtilityAssociationsPopupElement -> {
+                    states.add(
+                        element,
+                        UtilityAssociationsElementState(element, coroutineScope)
+                    )
+                }
+
+                else -> {
+                    // TODO remove for release
+                    println("encountered element of type ${element::class.java}")
+                }
+            }
+        }
+
+        return states
+    }
+}
+
+/**
+ * A structure that holds the [Popup] and its associated [PopupElementStateCollection].
+ *
+ * This class is also [Stable] and enables composition optimizations.
+ *
+ * @param popup the [Popup] to create the state for.
+ * @param stateCollection the [PopupElementStateCollection] created for the [Popup].
+ */
+@Stable
+internal data class PopupStateData(
+    val popup: Popup,
+) {
+    internal lateinit var stateCollection: PopupElementStateCollection
+        private set
+    /**
+     * Indicates if the evaluateExpression function for the [popup] has been run.
+     */
+    internal var initialEvaluation : MutableState<Boolean> = mutableStateOf(false)
+        private set
```

Comment:
```suggestion
    var initialEvaluation : Boolean by mutableStateOf(false)
        private set
```
This needs to be a var of type `Boolean` for the private setter to work. Otherwise any consumer can simply set a value directly like so -
```kotlin
initialEvaluation.value = false
```

I also don't think there is a need to mark these properties and methods as internal since the class is already marked internal.

Similarly for other properties below too.

Reply:
Updated. We should make similar change for FormStateData as well 

https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/blob/feature-branches/forms/toolkit/featureforms/src/main/java/com/arcgismaps/toolkit/featureforms/FeatureFormState.kt#L331